Implement Kruskal’s Minimum Spanning Tree Algorithm

Hi @manparvesh I'm trying to build and test however the build script is returning:
```
Scanning dependencies of target googletest
make[2]: *** No rule to make target '../external/googletest/googletest/src/gtest-all.cc', needed by 'CMakeFiles/googletest.dir/external/googletest/googletest/src/gtest-all.cc.o'.  Stop.
CMakeFiles/Makefile2:686: recipe for target 'CMakeFiles/googletest.dir/all' failed
make[1]: *** [CMakeFiles/googletest.dir/all] Error 2
Makefile:94: recipe for target 'all' failed
make: *** [all] Error 2
make failed
```
I've never used googletest before so I'm not sure how to fix this, can you help me out here?

--- a/src/graph/kruskal_minimum_spanning_tree.hpp
+++ b/src/graph/kruskal_minimum_spanning_tree.hpp
@@ -1,0 +1,86 @@
+//
+// Created by brand-temp on 8/10/2018.
+//
+
+// Whole thing needs a redesign
+// return vector of edges
+// return total weight and vecotr of edges?
+// yikes...
+// Ideas.....
+/*
+1. send in pointer, return weight of mst, and the thing at the pointer is changed to the tree
+2. return pair of totaly weight then edges
+*/
+
+#include <iostream>
+#include <vector>
+
+using namespace std;
+
+
+/**
+ * kruskal_minimum_spanning_tree - implements the Kruskal algorithm
+ * @tparam T - The label type for each vertex in the graph, i.e. int or string
+ * @param v - vector of Ts representing each vertex in the graph
+ * @param e - vector of pairs of integers and T pairs representing each edge in the graph
+ * @return a vector of edges and their weights
+ */
+
+template<class T>
+struct edge {
+    T start_point;
+    T end_point;
+    int weight;
+    edge(T s, T e, int w) : start_point(s), end_point(e), weight(w) { }
+    bool operator<(const edge &rhs) const {
+        return weight < rhs.weight;
+    }
+};
+template<class T>
+pair<int, vector<edge<T>>> kruskal_minimum_spanning_tree(vector<T> v, vector<edge<T>> e) {
+    // make vector of "trees"
+    vector<vector<T>> trees;
+    vector<edge<T>> mst;
+    int weight = 0;
+    for(T vertex : v) {
+        vector<T> tree;
+        tree.push_back(vertex);
+        trees.push_back(tree);
+    }
+    sort(e.rbegin(), e.rend());
+    while(e.size() > 0) {
+        edge<T> curr = e.back();
+        e.pop_back();
+
+        T a = curr.start_point, b = curr.end_point;
+        int a_index = -1, b_index = -1;
+        for(int i = 0; i < trees.size(); i++) {
+            vector<T> tree = trees[i];
+            for(T label : tree) {
+                if (label == a)
+                    a_index = i;
+                if (label == b)
+                    b_index = i;
+            }
+        }
+        if (a_index == -1 || b_index == -1)
+            continue;
+        if(a_index == b_index)
+            continue;
+        weight += curr.weight;
+        mst.push_back(curr);
+
+        trees[a_index].insert(trees[a_index].end(), trees[b_index].begin(), trees[b_index].end());
+        trees.erase(trees.begin() + b_index);
+    }
+    pair<int, vector<edge<T>>> toRet;
+    if(trees.size() > 1) {
+        toRet.first = 0;
+        cout << "This graph doesn't appear to be fully connected. All edges have been exhausted, "
+                "but there are still unconnected vertices." << endl;
+        return toRet;
+    }
+    toRet.first = weight;
+    toRet.second = mst;
+    return toRet;
+}

--- a/src/graph/kruskal_minimum_spanning_tree.hpp
+++ b/src/graph/kruskal_minimum_spanning_tree.hpp
@@ -2,30 +2,15 @@
 // Created by brand-temp on 8/10/2018.
 //
 
-// Whole thing needs a redesign
-// return vector of edges
-// return total weight and vecotr of edges?
-// yikes...
-// Ideas.....
-/*
-1. send in pointer, return weight of mst, and the thing at the pointer is changed to the tree
-2. return pair of totaly weight then edges
-*/
-
 #include <iostream>
 #include <vector>
 
 using namespace std;
 
-
 /**
- * kruskal_minimum_spanning_tree - implements the Kruskal algorithm
- * @tparam T - The label type for each vertex in the graph, i.e. int or string
- * @param v - vector of Ts representing each vertex in the graph
- * @param e - vector of pairs of integers and T pairs representing each edge in the graph
- * @return a vector of edges and their weights
+ *
+ * @tparam T - The type of label used for each edge
  */
-
 template<class T>
 struct edge {
     T start_point;
@@ -36,6 +21,15 @@ struct edge {
         return weight < rhs.weight;
     }
 };
+
+/**
+ *
+ * @tparam T - The type of the label for your vertices - e.g. char or string.
+ * @param v - A vector which names each vertex in the original graph
+ * @param e - A vector of edges in the graph
+ * @return a pair which stores the final total weight of the MST, and the edges of the MST
+ */
+
 template<class T>
 pair<int, vector<edge<T>>> kruskal_minimum_spanning_tree(vector<T> v, vector<edge<T>> e) {
     // make vector of "trees"

--- a/test/graph/kruskal_minimum_spanning_tree.cpp
+++ b/test/graph/kruskal_minimum_spanning_tree.cpp
@@ -1,0 +1,83 @@
+#include "gtest/gtest.h"
+#include "kruskal_minimum_spanning_tree.hpp"
+#include <iostream>
+
+using namespace std;
+
+TEST(kruskal_minimum_spanning_tree, all_tests) {
+    vector<char> vertex;
+    vertex.push_back('A');
+    vertex.push_back('B');
+    vertex.push_back('C');
+    vertex.push_back('D');
+    vertex.push_back('E');
+
+    vector<edge<char>> edges;
+    edges.push_back(edge<char>('A', 'B', 1));
+    edges.push_back(edge<char>('A', 'C', 7));
+    edges.push_back(edge<char>('A', 'D', 6));
+    edges.push_back(edge<char>('B', 'E', 4));
+    edges.push_back(edge<char>('C', 'E', 1));
+    edges.push_back(edge<char>('D', 'E', 1));
+
+
+
+    pair<int, vector<edge<char>>> returnTree = kruskal_minimum_spanning_tree(vertex, edges);
+
+    cout << "-------------------\nKRUSKAL 1\n-------------------" << endl;
+    cout << "Expected weight was 7, actual weight was: " << returnTree.first << endl;
+    cout << "Edges were: " << endl;
+    for(edge<char> e: returnTree.second) {
+        cout << e.start_point << " to " << e.end_point << " with weight: " << e.weight << endl;
+    }
+
+    ASSERT_TRUE(returnTree.first == 7);
+
+    vertex.push_back('F');
+    vertex.push_back('G');
+
+    edges.push_back(edge<char>('G', 'D', 3));
+    edges.push_back(edge<char>('G', 'A', 4));
+    edges.push_back(edge<char>('F', 'C', 2));
+    edges.push_back(edge<char>('F', 'G', 1));
+
+    returnTree = kruskal_minimum_spanning_tree(vertex, edges);
+
+    cout << "-------------------\nKRUSKAL 2\n-------------------" << endl;
+    cout << "Expected weight was 10, actual weight was: " << returnTree.first << endl;
+    cout << "Edges were: " << endl;
+    for(edge<char> e: returnTree.second) {
+        cout << e.start_point << " to " << e.end_point << " with weight: " << e.weight << endl;
+    }
+
+    ASSERT_TRUE(returnTree.first == 10);
+
+    returnTree = kruskal_minimum_spanning_tree(vertex, edges);
+
+    edges.pop_back();
+    edges.push_back(edge<char>('D', 'B', -10));
+
+    returnTree = kruskal_minimum_spanning_tree(vertex, edges);
+
+    cout << "-------------------\nKRUSKAL 3\n-------------------" << endl;
+    cout << "Expected weight was -2, actual weight was: " << returnTree.first << endl;
+    cout << "Edges were: " << endl;
+    for(edge<char> e: returnTree.second) {
+        cout << e.start_point << " to " << e.end_point << " with weight: " << e.weight << endl;
+    }
+
+    ASSERT_TRUE(returnTree.first == -2);
+
+    vertex.push_back('H');
+
+    returnTree = kruskal_minimum_spanning_tree(vertex, edges);
+
+    cout << "-------------------\nKRUSKAL 4\n-------------------" << endl;
+    cout << "Expected weight was 0, actual weight was: " << returnTree.first << endl;
+    cout << "Edges were: " << endl;
+    for(edge<char> e: returnTree.second) {
+        cout << e.start_point << " to " << e.end_point << " with weight: " << e.weight << endl;
+    }
+
+    ASSERT_TRUE(returnTree.first == 0 && returnTree.second.size() == 0);
+}

